### ISSUE TITLE
CWE-400 | Uncontrolled resource consumption

### ISSUE THREAD

**Issue by archanasharma3** [micromatch/braces#35]
Ran latest scan and landed with this issue - CWE-400 | Uncontrolled resource consumption

The NPM package "braces" fails to limit the number of characters it can handle, which could lead to Memory Exhaustion. In "lib/parse.js," if a malicious user sends "imbalanced braces" as input, the parsing will enter a loop, which will cause the program to start allocating heap memory without freeing it at any moment of the loop. Eventually, the JavaScript heap limit is reached, and the program will crash.

 braces/lib/parse.js on line 308  - do {

**Comment by doowb** [comment]
@archanasharma3 could you provide the error you're seeing when the heap limit is reached and your program crashes?

The pattern length is checked [here](https://github.com/micromatch/braces/blob/98414f9f1fabe021736e26836d8306d5de747e0d/lib/parse.js#L37-L40).
You could either check the pattern length before passing it to `braces` or handle the error in a try-catch.

There's also [an option](https://github.com/micromatch/braces?tab=readme-ov-file#optionsmaxlength) that will enforce a lower max length if you'd like to do that and catch the error.

**Comment by rahulmukherjee85** [comment]
In addition to @archanasharma3 comments,
braces flagged as vulnerable package:
Dependency npm:braces 3.0.2 is vulnerable
Cxab55612e-3a56 7.5 - CWE-400
Results powered by Checkmarx(c)
Braces is a transitive dependency of Npm-chokidar in our application.
there any way to set the limit in global level ? please let us know

**Comment by MarioTeixeiraCx** [comment]
Hello @doowb,

I'm reaching out here to confirm that this is indeed the other vulnerability reported by the Checkmarx CxResearch group. We opened this issue (https://github.com/micromatch/micromatch/issues/243) after trying to reach out by email unsuccessfully. You can still find the details in the report we sent.

CVEs for both vulnerabilities will come out soon.

Thank you,
Mário Teixeira

**Comment by LoganArnett** [comment]
It looks like the description of this vulnerability is around the do/while loop for the imbalanced braces:
```
The NPM package "braces" fails to limit the number of characters it can handle, which could lead to Memory Exhaustion. In "lib/parse.js," if a malicious user sends "imbalanced braces" as input, the parsing will enter a loop, which will cause the program to start allocating heap memory without freeing it at any moment of the loop. Eventually, the JavaScript heap limit is reached, and the program will crash.
```

**Comment by MarioTeixeiraCx** [comment]
Hello @doowb ,

CVEs for `micromatch` and `braces` vulnerabilities are now public. You can see this comment for additional details:
https://github.com/micromatch/micromatch/issues/243#issuecomment-2107198254

Best regards,
Mário Teixeira

**Comment by sdupont73** [comment]
Does anyone know when a fix is expected for this vulnerability?

**Comment by benjsmi** [comment]
@sdupont73 -- I'm going to guess that the answer to "when is a fix coming" is probably **never**. The last commit to this repository was 2019.  Same goes for the npm.  I do not maintain this project, so I don't actually **know**, but we've not heard anything from the maintainers and there seems to be very little development effort, so my guess is that a replacement is a better option.

**Comment by BlueTux611** [comment]
For this issue proposed [this](https://github.com/micromatch/braces/pull/38) as potential fix:

**Comment by prabhu** [comment]
So you found an infinite loop and then decided to call it a security vulnerability? The amount of time human civilization wastes with low quality CVEs is staggering.

**Comment by theta682** [comment]
I see a commit 9f5b4cf47329351bcb64287223ffb6ecc9a5e6d3 which is included in v3.0.3. Probably this means the issue is fixed.
